numpy/core/src/multiarray/descriptor.c

```diff
@@ -1017,19 +1007,21 @@ _convert_from_dict(PyObject *obj, int align)
     /*
      * Use PyMapping_GetItemString to support dictproxy objects as well.
      */
-    names = Borrowed_PyMapping_GetItemString(obj, "names");
-    descrs = Borrowed_PyMapping_GetItemString(obj, "formats");
+    names = PyMapping_GetItemString(obj, "names");
+    descrs = PyMapping_GetItemString(obj, "formats");
```

Comment:
Pretty sure it's not safe to do these both before checking if either errored. Consider a pathological type like:
```
class BadDict:
    def __getitem__(self, item):
        raise TypeError("Exception in flight here causes python to be unhappy")

class Foo:
    ___array_interface__ = BadDict
```

Note that we need a type overloading `__getitem__` like this to make your patch worth it in the first place.

Reply:
Fixing, even though in this case I am pretty sure the code will work correctly, both `name` and `descrs` will be `NULL` so the following `if` will kick in. 

The patch is to ensure we do not use an object with refcount 0, since it could be collected at any time.

Reply:
Looking more carefully, the code clears any exceptions before calling `_use_fields_dict` so is there a problem with the test you propose?

I will add a pathological test that only errors on `name` lookup

Reply:
For reference, the error code path is hit when a `dict` with no `name`, `format`. For instance when converting this valid spec

```
dt = np.dtype({'f0': ('i4', 0), 'f1':('u1', 4)}, align=True)
```

Reply:
I've seen problems before where `__getitem__("formats")` would raise a `SystemError` because an exception was set by `__getitem__("names")`. I think in this case you silence both anyway, but in principle you could hit an assert or worse.

Reply:
If that is something we should be checking for, we should only silence a `KeyError`. Would you like that to be part of this PR?

Reply:
That's a fair point, but I think it's fine to leave for another PR.

Reply:
So to clarify - I'd like to see this changed to call one, check for null, then call the other.  That's in line with the purpose of this PR, which is to make this work for non-`dict` objects with `__getitem__` (the only case when the refcount could drop to 0)

Catching `KeyError` specifically is a nice idea, but not one that really makes sense as part of this PR.

Reply:
got it, thanks for persevering. Fixed.